### PR TITLE
Document nftables kube-proxy alpha (KEP 3866)

### PR DESCRIPTION
**Pull request by danwinship**
Enhancement: https://github.com/kubernetes/enhancements/issues/3866
KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/3866-nftables-proxy/README.md
Implementation: https://github.com/kubernetes/kubernetes/pull/121046

### CODE REVIEW

**Comment by sftim** [comment]
/retitle [WIP] Document nftables kube-proxy alpha (KEP 3866)

**Comment by harshitasao** [comment]
Hi @danwinship , Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before **Tuesday 14th November 2023**.
Thank you!

**Comment by danwinship** [comment]
I'm not sure about how to document this... I know we're not supposed to document in reference to specific releases, but I think it's important to be clear that in 1.29, this is not necessarily expected to be better than iptables. I expect this entire section will be rewritten for 1.30 (when we will still be alpha, but it will be a _better_ alpha).

**Comment by sftim** [comment]
> I'm not sure about how to document this... I know we're not supposed to document in reference to specific releases, but I think it's important to be clear that in 1.29, this is not necessarily expected to be better than iptables. I expect this entire section will be rewritten for 1.30 (when we will still be alpha, but it will be a _better_ alpha).

Aim to write docs as if the improvements slated for v1.30 probably won't happen, but still might.
You can add eg
```markdown
For kube-proxy {{< skew currentVersion >}} on Linux nodes, the Kubernetes project recommends using either
`ipvs` or `iptables` proxy mode.
```
and then drop that sentence whenever the situation changes. Even if the improvements get delayed to v1.31, the v1.29 and v1.30 docs won't mislead readers.

File: content/en/docs/reference/networking/virtual-ips.md [hunk 1]
```diff
@@ -268,6 +271,24 @@ falls back to running in iptables proxy mode.
...
+{{< feature-state for_k8s_version="v1.29" state="alpha" >}}
+
+_This proxy mode is only available on Linux nodes._
+
+The `nftables` proxy mode is similar in most ways to [the `iptables`
+mode](#proxy-mode-iptables), but uses the newer nftables kernel API
+rather than iptables. Currently the only major difference in
+functionality is that the `iptables` mode supports connecting to
+[NodePort Services](/docs/concepts/services-networking/service/#type-nodeport)
+on `127.0.0.1`, but the `nftables` mode does not.
```
Comment (sftim) [inline content/en/docs/reference/networking/virtual-ips.md:285]: Try to describe this in a way that is valid even if we dropped the `iptables` proxy mode and it wasn't mentioned in documentation (or maybe just a stub to tell people it's no longer available).

File: content/en/docs/reference/networking/virtual-ips.md [hunk 2]
```diff
@@ -268,6 +271,24 @@ falls back to running in iptables proxy mode.
...
+
+Although the nftables API is designed to provide better performance
+and scalability than iptables, the kube-proxy `nftables` mode is still
+under heavy development and is not necessarily expected to outperform
+the `iptables` mode at this time.
```
Comment (sftim) [inline content/en/docs/reference/networking/virtual-ips.md:290]: Try something like this (feel free to correct the technical details):
```markdown
In this mode, kube-proxy watches the Kubernetes API for the addition, removal or modification
of Service and EndpointSlice objects.
The kube-proxy sets up shared netfilter tables and chains for container packet forwarding.
For each Service, the kube-proxy also sets up specific nftable rules that capture traffic to any Service
operating in node port mode. The rules send traffic to a valid backing endpoint for that Service
if one is available, using round-robin distribution.

Packet forwarding rules get removed as soon as the backing Pod disappears from the
associated EndpointSlice.
```

**Comment by harshitasao** [comment]
Hi, @danwinship ! Kubernetes v1.29 Docs shadow here. **Just a reminder that the deadline to have this PR reviewed and merged is Tuesday, 28 November**. Let me know if you need any help!

**Comment by katcosgrove** [comment]
Hello @danwinship! The deadline to have this PR merged is Tuesday, November 28. Enhancements that do not include required documentation may not be included in the release. Is there anything I can do to help get this ready to merge? Thanks!

**Comment by danwinship** [comment]
> Try to describe this in a way that is valid even if we dropped the `iptables` proxy mode

OK, well, the obvious way to say that the nftables mode works like the iptables mode, without referencing the iptables mode, is to just copy the iptables docs over to the nftables section. Except that led to some obvious refactoring...

**Comment by danwinship** [comment]
(I suppose we could merge just the last commit here and then debate the rest on `main` later.)

**Comment by danwinship** [comment]
> (I suppose we could merge just the last commit here and then debate the rest on `main` later.)

done

**Review by sftim — COMMENTED** [review]
/lgtm
/approve

Thank you for this work. This PR contains the germ of the end user experience around this new feature, and it's written to meet that user need.

Please do consider opening a new PR, against _dev-1.29_, that cherry picks the other commits.
Worst case, you change the base branch to _main_ in a few weeks, rebase, and we continue the discussion there.

**Pull request merged**